csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs

```diff
@@ -731,6 +731,109 @@ private void TestModelInputBOOL()
             }
         }
 
+        [Fact]
+        private void TestReusingRunOutputNonStringType()
+        {
+            // model takes 1x5 input of fixed type, echoes back
+            string modelPath = Path.Combine(Directory.GetCurrentDirectory(), "test_types_BOOL.pb");
+            using (var session = new InferenceSession(modelPath))
+            {
+                var container = new List<NamedOnnxValue>();
+                var tensorIn = new DenseTensor<bool>(new bool[] { true, false, true, false, true }, new int[] { 1, 5 });
+                var nov = NamedOnnxValue.CreateFromTensor("input", tensorIn);
+                container.Add(nov);
+                var res1 = session.Run(container);
+
+                // change the name of the DisposableNamedOnnxValue
+                res1.First().Name = "input";
+
+                // Run inferencing 2 times using the output of the first Run()
+                for(int i=0; i<2; ++i)
+                { 
+                    using (var res2 = session.Run(res1))
+                    {
+                        var tensorOut = res2.First().AsTensor<bool>();
+                        Assert.True(tensorOut.SequenceEqual(tensorIn));
+                    }
+                }
+
+                // Dispose the result tensor
+                res1.First().Dispose();
+
+                bool succeeded = false;
+
+                // Now try using the disposed output as input to another Run()
+                try
+                {
+                    // Run() should fail with a user friendly error message.
+                    session.Run(res1);
+                }
+
+                catch(ObjectDisposedException e)
+                {
+                    var errorString = "This instance of DisposableNamedOnnxValue has already been disposed";
+
+                    Assert.True(e.Message.Contains(errorString));
+
+                    succeeded = true;
+                }
+
+                Assert.True(succeeded);
+            }
+        }
```

Comment:
This test is repeated below. Since it's testing a different functionality (i.e. friendly error message on re-dispose), can we factor it out as a separate test (and hence avoid duplication)?

Reply:
There are 2 tests - one for non-string type and one for string type (since string is special-cased a lot). Within each test there are 2 sub-tests encapsulated - one passing the output of one Run() to another (main functionality) and the other to make sure the c# layer gracefully handles cases when the DisposableNamedOnnxValue has been disposed. As such I feel this encapsulation structure is okay and prefer to keep it as such, unless you feel strongly about splitting it out...

Reply:
If a  Xunit test fails in a CI run, there's no way to tell which of the 2 Assert()s in the function body caused the overall test failure. Creating a new Fact testing the re-dispose-exception separately would isolate  errors in a more granular way. Also, the de-duping should reduce the code footprint.  For this PR, I don't think separate  tests for string-vs-non-string  are needed ... one should be sufficient.

Reply:
Sure - I ll make the split then.

Reply:
I split up the tests into separate tests